README.md
# WiremockClient

[![Version](https://img.shields.io/cocoapods/v/WiremockClient.svg?style=flat)](http://cocoapods.org/pods/WiremockClient)
[![License](https://img.shields.io/cocoapods/l/WiremockClient.svg?style=flat)](http://cocoapods.org/pods/WiremockClient)
[![Platform](https://img.shields.io/cocoapods/p/WiremockClient.svg?style=flat)](http://cocoapods.org/pods/WiremockClient)

WiremockClient is an HTTP client that allows users to interact with a standalone Wiremock instance from within an Xcode project.

## Installation

WiremockClient is available through [CocoaPods](http://cocoapods.org) and [Swift Package Manager](https://swift.org/package-manager/).

To install with CocoaPods, simply add the following line to your Podfile:

```ruby
pod "WiremockClient"
```

To install with SwiftPM, add the following line to the `dependencies` section of your Package.swift:

```swift
.package(url: "https://github.com/mobileforming/WiremockClient", .upToNextMajor(from: Version(major: 2, minor: 2, patch: 0)))
```

## Usage

WiremockClient maps closely to the functionality available in Wiremock's native Java API as described in the [documentation](http://wiremock.org/docs/). The pod enables you to build and post JSON mappings to a standalone Wiremock instance from within an Xcode project. It is assumed that you are familiar with the basics of using Wiremock, including initializing a standalone instance and populating it with mappings.

### Getting Started

To begin using WiremockClient, start up a standalone Wiremock instance on localhost port 8080. The base URL for all requests is set to `http://localhost:8080` by default and can be modified at the top of the `WiremockClient` file. Be sure to [whitelist your base URL](https://developer.apple.com/library/content/documentation/General/Reference/InfoPlistKeyReference/Articles/CocoaKeys.html#//apple_ref/doc/uid/TP40009251-SW33) before using WiremockClient, or you will be unable to communicate with your standalone instance.

### Posting a Mapping

The following code will post a mapping to your Wiremock standalone instance that will match any request sent to the `http://localhost:8080/my/path` endpoint and return a status of 200:

```swift
WiremockClient.postMapping(stubMapping:
    StubMapping.stubFor(requestMethod: .ANY, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .willReturn(
            ResponseDefinition()
                .withStatus(200)
    )
)
```

This behavior maps closely to the Java API as described in the [Stubbing](http://wiremock.org/docs/stubbing/) portion of the documentation, so a full explanation will not be reproduced here. There are three significant differences to note:
1. The `stubFor()` method used to initialize a mapping in the Java API is now a type function of the `StubMapping` class.
2. The `aResponse()` method used to initialize a response object in the Java API has been replaced with an instance of the `ResponseDefinition` class.
3. The mapping created using the `stubFor` and `aResponse` methods above must be passed to the `WiremockClient.postMapping(stubMapping: StubMapping)` function to be posted to the standalone Wiremock instance.

### Matching a Request

All of the request matching logic available in the Java API has been reproduced within WiremockClient. For a full explanation of the methods below, reference the [Request Matching](http://wiremock.org/docs/request-matching/) portion of the documentation.

A collection of `StubMapping` instance methods enable the user to build mappings step-by-step, specifying the criteria that must be met in order for an incoming network request to be considered a match:

```swift
WiremockClient.postMapping(stubMapping:
    StubMapping.stubFor(requestMethod: .ANY, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .withHeader("Accept", matchCondition: .contains, value: "xml")
        .withCookie("session", matchCondition: .matches, value: ".*12345.*")
        .withQueryParam("search_term", matchCondition: .equalTo, value: "WireMock")
        .withBasicAuth(username: "myUsername", password: "myPassword")
        .withRequestBody(.equalTo, value: "Some request body string")
        .willReturn(
	    ResponseDefinition()
    )
)
```

An additional `withRequestBodyEqualToJson` method has been added to allow users to set the `ignoreArrayOrder` and `ignoreExtraElements` flags described in the ‘JSON equality’ section of the [Request Matching](http://wiremock.org/docs/request-matching/) documentation:

```swift
WiremockClient.postMapping(stubMapping:
    StubMapping.stubFor(requestMethod: .ANY, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .withRequestBodyEqualToJson(jsonString: "{ \"total_results\": 4 }", ignoreArrayOrder: true, ignoreExtraElements: true)
        .willReturn(ResponseDefinition())
)
```

WiremockClient also includes a convenience method for stubbing request JSON that is stored in a local file:

```swift
WiremockClient.postMapping(stubMapping:
    StubMapping.stubFor(requestMethod: .ANY, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .withRequestBodyFromLocalJsonFile(fileName: "myFile", in: Bundle(for: type(of: self)))
        .willReturn(ResponseDefinition())
)
```

Mappings can also be prioritized as described in the ‘Stub priority’ section of the [Stubbing](http://wiremock.org/docs/stubbing/) documentation:

```swift
WiremockClient.postMapping(stubMapping:
    StubMapping.stubFor(requestMethod: .ANY, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .withPriority(1)
        .willReturn(ResponseDefinition())
)
```

### Defining a Response

All of the response definition logic available in the Java API has been reproduced in WiremockClient. For a full explanation of the methods below, reference the [Stubbing](http://wiremock.org/docs/stubbing/) portion of the documentation.

A collection of `ResponseDefinition` instance methods enable the user to specify elements to include in the response that is returned when a mapping is matched:

```swift
WiremockClient.postMapping(stubMapping:
    StubMapping.stubFor(requestMethod: .ANY, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .willReturn(
            ResponseDefinition()
                .withStatus(200)
                .withStatusMessage("Great jorb!")
                .withHeaders(["Pragma": "no-cache", "Connection": "keep-alive"])
                .withBody("Just a plain old text body")
    )
)
```

WiremockClient also includes a convenience method for returning JSON that is stored in a local file:

```swift
WiremockClient.postMapping(stubMapping:
    StubMapping.stubFor(requestMethod: .ANY, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .willReturn(
            ResponseDefinition()
                .withLocalJsonBodyFile(fileName: "myFile", in: Bundle(for: type(of: self)))
    )
)
```

### Proxying

As in the Java API, requests can be proxied through to other hosts. A full explanation of this method can be found in the [Proxying](http://wiremock.org/docs/proxying/) portion of the documentation:

```swift
WiremockClient.postMapping(stubMapping:
    StubMapping.stubFor(requestMethod: .ANY, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .willReturn(
            ResponseDefinition()
                .proxiedFrom("http://myproxyhost.gov")
    )
)
```

### Stateful Behavior

WiremockClient also supports scenarios as described in the [Stateful Behavior](http://wiremock.org/docs/stateful-behaviour/) portion of the documentation:

```swift
WiremockClient.postMapping(stubMapping:
    StubMapping.stubFor(requestMethod: .GET, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .inScenario("Scenario Title")
        .whenScenarioStateIs("Required Scenario State")
        .willSetStateTo("New Scenario State")
        .willReturn(
            ResponseDefinition()
            .withStatus(200)
    )
)
```

The following method resets all scenarios to their default state (“Started”):

```swift
WiremockClient.resetAllScenarios()
```

### Updating a Mapping

Updating a mapping requires a reference to it’s UUID. When a mapping is created, a UUID is automatically assigned to it. However, it is also possible to assign a UUID manually and cache it in a variable for future reference. In the example below, a mapping is posted that returns a status code of 200 when matched. The mapping is then updated to return a status code of 404:

```swift
let myMappingID = UUID()

WiremockClient.postMapping(stubMapping:
    StubMapping.stubFor(requestMethod: .GET, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .withUUID(myMappingID)
        .willReturn(
            ResponseDefinition()
            .withStatus(200)
    )
)

WiremockClient.updateMapping(uuid: myMappingID, stubMapping:
    StubMapping.stubFor(requestMethod: .GET, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .willReturn(
            ResponseDefinition()
            .withStatus(404)
    )
)
```

### Deleting Mappings

Similar to updating a mapping, deleting a mapping requires a reference to it’s UUID:

```swift
let myMappingID = UUID()

WiremockClient.postMapping(stubMapping:
    StubMapping.stubFor(requestMethod: .GET, urlMatchCondition: .urlEqualTo, url: "http://localhost:8080/my/path")
        .withUUID(myMappingID)
        .willReturn(
            ResponseDefinition()
            .withStatus(200)
    )
)

WiremockClient.deleteMapping(uuid: myMappingID)
```

It is also possible to reset your Wiremock instance by deleting all mappings simultaneously:

```swift
WiremockClient.reset()
```

### Saving Mappings

Mappings can be persisted to the `mappings` directory of your Wiremock instance via the following method:

```swift
WiremockClient.saveAllMappings()
```

### Using WiremockClient in an XCTestCase

A typical use case of WiremockClient looks like this:
1. Call `WiremockClient.postMapping()` in the test suite’s `setup()` method to post the required mappings before the app launches.
2. If necessary, call `WiremockClient.updateMapping()` within the test script to alter mappings on the fly.
3. Call `WiremockClient.reset()` in the test suite’s `tearDown()` method to remove all mappings after the test has finished.

## Author

Ted Rothrock, theodore.rothrock@gmail.com

## License

WiremockClient is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 3.0.0 - 2023-02-14
### Added
- Support for Swift Package Manager.
- StubMapping `withRequestBodyFromLocalJsonFile` method to support defining mappings via local JSON files.
### Changed
- Bumped iOS deployment target to 14.0.
- Bumped macOS deployment target to 11.0.

## 2.1.0 - 2022-05-26
### Changed
- Adjusted access levels to allow stub and mapping properties to be read externally.

## 2.0.0 - 2021-03-21
### Added
- Support for adding a delay to individual responses.
- Support for deleting request history.
- Unit test coverage.
### Removed
- Public `json` and `data` properties from ResponseDefinition.
### Changed
- Public WiremockClient methods now throw errors.
- ResponseDefinition `withLocalJsonBodyFile` method updated to accept a Bundle reference instead of a `String`.

## 1.4.0 - 2020-09-04
### Added
- Support for adding a global delay to all Wiremock server responses
- Documentation to all public classes and methods

## 1.3.0 - 2020-02-11
### Added
- CHANGELOG.md
- Supported Swift versions in podspec (4.0, 5.0)
- Support for transformers in ResponseDefinition
- Minimum deployment target for tvOS to podspec
- Support for verifying requests made to the Wiremock server
### Removed
- .swift-version file


## 1.2.1 - 2018-06-15
### Changed
- Made all URLSession data tasks synchronous

## 1.2.0 - 2018-02-26
### Added
- Minimum deployment target for macOS to podspec
- Support for shutting down server Wiremock instance
- Support for checking if Wiremock instance is running
- Access to ResponseDefinition data
- Support for SPM

## 1.1.2 - 2017-10-19
### Changed
- .swift-version file syntax
- README
### Added
- Support for handling JSON array objects in ResponseDefinition

## 1.1.1 - 2017-10-12
### Added
- Support for adding headers to ResponseDefinition

## 1.1.0 - 2017-09-11
### Added
- Access to ResponseDefinition JSON

## 1.0.0 - 2017-07-26
### Added
- WiremockClient 1.0.0


Sources/WiremockClient/RequestMapping.swift
//
//  RequestMapping.swift
//  WiremockClientPackageDescription
//
//  Created by Tony Eichelberger on 10/30/18.
//

import Foundation

//----------------------------------
// MARK: Codable Request objects
//----------------------------------

/**
The example json from wiremock.org when only requests are returned.
 This is associated with the verify method:
 http://wiremock.org/docs/verifying
 
"url": "/my/other/url",
"absoluteUrl": "http://my.other.domain.com/my/other/url",
"method": "POST",
"headers": {
    "Accept": "text/plain",
    "Content-Type": "text/plain"
},
"body": "My text",
"browserProxyRequest": false,
"loggedDate": 1339083581823,
"loggedDateString": "2012-06-07 16:39:41"

 */

internal struct AllLoggedRequests: Codable {
    public var requests: [LoggedRequest]
}

public struct LoggedRequest: Codable {
    public var url: String?
    public var absoluteUrl: String?
    public var method: RequestMethod?
    public var body: String?
    public var browserProxyRequest: Bool?
    public var loggedDateString: String?
    public var headers: [String : String]?
}

/// An object used to configure a Wiremock request verification mapping. Refer to http://http://wiremock.org/docs/verifying/ for more details.
public class RequestMapping {
    
    private var request: RequestPattern
    
    private init(requestMethod: RequestMethod, urlMatchCondition: URLMatchCondition, url: String) {
        self.request = RequestPattern(requestMethod: requestMethod, urlMatchCondition: urlMatchCondition, url: url)
    }
    
    //----------------------------------
    // MARK: Mapping Builder Methods
    //----------------------------------
    
    /// Initializes a `RequestMapping`
    ///
    /// - Parameter requestMethod: The HTTP request method to match
    /// - Parameter urlMatchCondition: A condition used to match the URL path and parameters of a request to a mapping
    /// - Parameter url: The URL path and parameter pattern that the mapping will use to match requests
    /// - Returns: A configured `RequestMapping`
    public static func requestFor(requestMethod: RequestMethod, urlMatchCondition: URLMatchCondition, url: String) -> RequestMapping {
        return RequestMapping(requestMethod: requestMethod, urlMatchCondition: urlMatchCondition, url: url)
    }
    
    /// Adds a header match condition to a mapping
    ///
    /// - Parameter key: The header key
    /// - Parameter matchCondition: The condition under which the header will be evaluated as a match
    /// - Parameter value: The header value
    /// - Returns: The `RequestMapping` with an updated header match condition
    public func withHeader(_ key: String, matchCondition: MatchCondition, value: String) -> RequestMapping {
        self.request.headers = self.request.headers ?? [String: [String: String]]()
        let headerDict = [matchCondition.rawValue: value]
        self.request.headers?[key] = headerDict
        return self
    }
    
    /// Adds a cookie match condition to a mapping
    ///
    /// - Parameter key: The cookie key
    /// - Parameter matchCondition: The condition under which the cookie will be evaluated as a match
    /// - Parameter value: The cookie value
    /// - Returns: The `RequestMapping` with an updated cookie match condition
    public func withCookie(_ key: String, matchCondition: MatchCondition, value: String) -> RequestMapping {
        self.request.cookies = self.request.cookies ?? [String: [String: String]]()
        let cookieDict = [matchCondition.rawValue: value]
        self.request.cookies?[key] = cookieDict
        return self
    }
    
    /// Adds a query parameter match condition to a mapping
    ///
    /// - Parameter param: The param key
    /// - Parameter matchCondition: The condition under which the param will be evaluated as a match
    /// - Parameter value: The param value
    /// - Returns: The `RequestMapping` with an updated param match condition
    public func withQueryParam(_ param: String, matchCondition: MatchCondition, value: String) -> RequestMapping {
        self.request.queryParameters = self.request.queryParameters ?? [String: [String:Any]]()
        let paramDict = [matchCondition.rawValue: value]
        self.request.queryParameters?[param] = paramDict
        return self
    }
    
    /// Adds a basic authentication match condition to a mapping
    ///
    /// - Parameter username: The username to evaluate
    /// - Parameter password: The password to evaluate
    /// - Returns: The `RequestMapping` with an updated basic authentication match condition
    public func withBasicAuth(username: String, password: String) -> RequestMapping {
        self.request.basicAuthCredentials = ["username": username, "password": password]
        return self
    }
    
    /// Adds a request body match condition to a mapping
    ///
    /// - Parameter matchCondition: The condition under which the body will be evaluated as a match
    /// - Parameter value: The body value
    /// - Returns: The `RequestMapping` with an updated body match condition
    public func withRequestBody(_ matchCondition: MatchCondition, value: String) -> RequestMapping {
        self.request.bodyPatterns = self.request.bodyPatterns ?? [[String: Any]]()
        let bodyPatternDict = [matchCondition.rawValue: value]
        self.request.bodyPatterns?.append(bodyPatternDict)
        return self
    }
    
    /// Adds a request JSON match condition to a mapping
    ///
    /// - Parameter jsonString: The request JSON value in `String` form
    /// - Parameter ignoreArrayOrder: A flag that indicates if matching JSON must contain elements in an exact order
    /// - Parameter ignoreExtraElements: A flag that indicates if matching JSON must not contain extra elements
    /// - Returns: The `RequestMapping` with an updated JSON match condition
    public func withRequestBodyEqualToJson(jsonString: String, ignoreArrayOrder: Bool, ignoreExtraElements: Bool) -> RequestMapping {
        self.request.bodyPatterns = self.request.bodyPatterns ?? [[String: Any]]()
        var bodyPatternDict: [String: Any] = [MatchCondition.equalToJson.rawValue: jsonString]
        if ignoreArrayOrder {
            bodyPatternDict["ignoreArrayOrder"] = true
        }
        if ignoreExtraElements {
            bodyPatternDict["ignoreExtraElements"] = true
        }
        self.request.bodyPatterns?.append(bodyPatternDict)
        return self
    }
    
    //----------------------------------
    // MARK: Mapping to Data Conversion
    //----------------------------------
    
    // Mapping Key Names
    
    let keyMethod           = "method"
    let keyParams           = "queryParameters"
    let keyHeaders          = "headers"
    let keyCookies          = "cookies"
    let keyBasicAuth        = "basicAuthCredentials"
    let keyBodyPatterns     = "bodyPatterns"
    
    internal func requestDict() -> [String: Any] {
        
        var requestDict = [String: Any]()
        
        // URL
        
        requestDict["\(self.request.urlPattern.urlMatchCondition.rawValue)"] = "\(self.request.urlPattern.url)"
        
        // Request Method
        
        requestDict[keyMethod] = "\(self.request.requestMethod.rawValue)"
        
        // Headers
        
        if let headers = self.request.headers {
            requestDict[keyHeaders] = headers
        }
        
        // Cookies
        
        if let cookies = self.request.cookies {
            requestDict[keyCookies] = cookies
        }
        
        // Query Parameters
        
        if let queryParameters = self.request.queryParameters {
            requestDict[keyParams] = queryParameters
        }
        
        // Basic Auth Credentials
        
        if let credentials = self.request.basicAuthCredentials {
            requestDict[keyBasicAuth] = credentials
        }
        
        // Request Body Patterns
        
        if let bodyPatterns = self.request.bodyPatterns {
            requestDict[keyBodyPatterns] = bodyPatterns
        }
        
        return requestDict
    }
    
    internal func asRequestData() -> Data? {
        return try? JSONSerialization.data(withJSONObject: requestDict(), options: [.prettyPrinted])
    }
    
}

extension RequestMapping: CustomStringConvertible {
    /// A `String` representation of the mapping
    public var description: String {
        guard let data = asRequestData(), let stringVal = String(data: data, encoding: .utf8) else {
            return "{}"
        }
        return stringVal
    }
}
